### PR TITLE
Fix action parameter validation when action overrides a parameter

### PR DESCRIPTION
**Pull request by Kami**
This pull request fixes a bug described here https://github.com/StackStorm/st2/pull/2335#issuecomment-169615849.

The problem was that we didn't perform a deep merge of runner + action parameters which means that if user would override runner parameter in an action and only specify a single parameter attribute (e.g. only immutable, but not type, etc.) things would explode.

Schema validation code changes didn't touch this part so probably the only reason it worked in the past is because we didn't actually perform full strict validation of the runner parameter types (lovely, eh? :P)

TODO:

- [x] Tests

### CODE REVIEW

File: st2common/st2common/util/misc.py [hunk 1]
```diff
@@ -92,3 +93,18 @@ def rstrip_last_char(input_str, char_to_strip):
         return input_str[:-len(char_to_strip)]
 
     return input_str
+
+
+def deep_update(d, u):
```
Comment (Kami) [inline st2common/st2common/util/misc.py]: I was looking for an existing library, but the one I found on PyPi actually breaks on some edge cases so I just decided to use a simple function.

**Comment by m4dcoder** [comment]
@Kami: I think this is duplicate of https://github.com/StackStorm/st2/pull/2359

**Comment by estee-tew** [comment]
https://stackstorm.slack.com/archives/stackstorm/p1452194722021258

**Comment by Kami** [comment]
@m4dcoder I think those are two different issues, at least that's what I thought when I first saw your PR.

**Comment by estee-tew** [comment]
https://stackstorm.slack.com/archives/stackstorm/p1452202169021366

**Comment by Kami** [comment]
It looks like I can't really resolve conflicts and merge this PR until we decide on this - https://github.com/StackStorm/st2/pull/2359#issuecomment-169921028 so I will wait on that.

**Comment by estee-tew** [comment]
https://stackstorm.slack.com/archives/stackstorm/p1452245563021539

**Comment by Kami** [comment]
Alright, I resolved the merge conflicts as discussed with @m4dcoder.

I will go ahead, merge this PR into master and open a new PR for auto-generated documentation for parameter attributes which can be overriden.

**Pull request merged**